### PR TITLE
Update Hurricane.js

### PR DESCRIPTION
**Pull request by Using-name**
Updated the entire thing to be even smarter.
In no particular order:
- Smart build placement
- Raider flash and retreat
- Bugfixes
- Better army decision making
- Calculates projected supply needed
- Proxies
- Better worker rush logic
- Refined surrender mechanic
- Disabled airships
- Added additional build orders
- Improved retreat logic
- Implemented value-based army logic (utility AI)
- Implemented scouting
- Implemented tower cheesing/usage
- Added a keepTogether function to keep armies together in a more condensed ball
- Added a automatic transition to mech after 2-3 bases are built
- Improved castle placement
- Kiting now uses the proper scope.setTimeout syntax and has the proper delay afterwards
- Kiting now will not happen if the army is in a corner
- Units in an army will now prioritize attacking units attacking other units, over attacking buildings

### CODE REVIEW

File: Hurricane.js [hunk 1]
```diff
@@ -161,13 +179,622 @@ class Randomizer {
...
+  return path;
+}
+
+function getHeap() {
+  return new BinaryHeap(function(node) {
+    return node.f;
+  });
+}
+
+function astar(){
```
Comment (sqrtofneg1) [inline Hurricane.js:203]: Astar, nice!

Comment (sqrtofneg1) [reply Hurricane.js:203]: even if it is copied :)

File: Hurricane.js [hunk 2]
```diff
@@ -731,351 +1656,278 @@ class RandBuild {
...
+        for(let x = 0; x < 10; x++){
+            for(let y = 0; y < 10; y++){
+                arr.push(scope.mapNodes.getNodeAt(x, y).isWall() == true ? 0 : 1);
+            }
         }
-        for(var i = 0; i < coordinates.length; i++){
-            this.tryTheseFirst.push(coordinates[i])
+        
+        //Ravaged
+        if(scope.numAllies == 0 && JSON.stringify(arr) == "[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,1,1,0,0,0,0,1,0,0,0,1,1,0,0,0,0,0,0,0,0,1,1,1,1]"){
```
Comment (sqrtofneg1) [inline Hurricane.js:1695]: hard coding for ravaged, makes sense

File: Hurricane.js [hunk 3]
```diff
@@ -2684,46 +4788,2471 @@ class RandChatter {
...
+                    unassignedRepairers.push(scope.mechRepairSquad[i]);
+                }//Because of JS magic, if you just use 
+                //let unassignedRepairers = scope.mechRepairSquad, unassignedRepairers will
+                //be an array of references to scope.mechRepairSquad (a shallow copy), kind
+                //of like an array of workers or soldiers or whatever. This means that
+                //if a change is made to unassignedRepairers without using the bit of code
+                //above, it will not only change unassignedRepairers, it will also change
+                //scope.mechRepairSquad.
+
+                //FYI, I spent over 4 hours trying to find this bug, so please appriciate
```
Comment (sqrtofneg1) [inline Hurricane.js:6310]: classic shallow copy vs deep copy headache

**Review by sqrtofneg1 — APPROVED** [review]

File: Hurricane.js (hunk 1 above)
Comment (Using-name) [reply Hurricane.js:203]: :)

File: Hurricane.js (hunk 3 above)
Comment (Using-name) [reply Hurricane.js:6310]: Yes!

File: Hurricane.js (hunk 2 above)
Comment (Using-name) [reply Hurricane.js:1695]: Easier than algorithm... I tried.

**Review by glaba — APPROVED** [review]

**Pull request merged**